File: net/OpenDraft.Lang/Message.cs
// <copyright file="Message.cs" company="Jon Rowlett">
// Copyright (c) Jon Rowlett. All rights reserved.
// </copyright>

namespace OpenDraft.Lang;

/// <summary>
/// Represents a message produced by any of the language components.
/// </summary>
public class Message
{
    /// <summary>
    /// Initializes a new instance of the <see cref="Message"/> class.
    /// </summary>
    /// <param name="severity">The severity of the message.</param>
    /// <param name="id">The ID of the message.</param>
    /// <param name="source">The source reference for the message.</param>
    /// <param name="text">The text of the message.</param>
    /// <exception cref="ArgumentNullException">Thrown if <paramref name="source"/> or <paramref name="text"/> is <c>null</c>.</exception>
    public Message(
        MessageSeverity severity,
        MessageId id,
        SourceReference source,
        string text)
    {
        this.Severity = severity;
        this.Id = id;
        this.Source = source ?? throw new ArgumentNullException(nameof(source));
        this.Text = text ?? throw new ArgumentNullException(nameof(text));
    }

    /// <summary>
    /// Gets the severity of this message.
    /// </summary>
    public MessageSeverity Severity { get; }

    /// <summary>
    /// Gets the ID of this message.
    /// </summary>
    public MessageId Id { get; }

    /// <summary>
    /// Gets the source reference for this message.
    /// </summary>
    public SourceReference Source { get; }

    /// <summary>
    /// Gets the text of the message.
    /// </summary>
    public string Text { get; }

    /// <inheritdoc/>
    public override string ToString()
    {
        return $"{this.Source}: {this.Severity} {this.Id}: {this.Text}";
    }
}

File: java/src/io/github/localtoast9001/opendraft/lang/Message.java
/**
 * Copyright (c) Jon Rowlett. All rights reserved.
 */
package io.github.localtoast9001.opendraft.lang;

/**
 * Represents a message generated during the processing of OpenDraft source code.
 */
public final class Message {
    private final SourceReference source;
    private final MessageId id;
    private final MessageSeverity severity;
    private final String text;

    /**
     * Constructs a Message with the specified source reference, message ID, severity, and text.
     * @param source the source reference of the message
     * @param id the message ID
     * @param severity the severity of the message
     * @param text the text of the message
     */
    public Message(
        final SourceReference source,
        final MessageId id,
        final MessageSeverity severity,
        final String text) {
        if (source == null) {
            throw new IllegalArgumentException("source cannot be null");
        }

        if (id == null) {
            throw new IllegalArgumentException("id cannot be null");
        }

        if (severity == null) {
            throw new IllegalArgumentException("severity cannot be null");
        }

        if (text == null) {
            throw new IllegalArgumentException("text cannot be null");
        }

        this.source = source;
        this.id = id;
        this.severity = severity;
        this.text = text;
    }

    /**
     * Returns the source reference of this message.
     * @return the source reference
     */
    public SourceReference getSource() {
        return source;
    }

    /**
     * Returns the message ID of this message.
     * @return the message ID
     */
    public MessageId getId() {
        return id;
    }

    /**
     * Returns the severity of this message.
     * @return the severity
     */
    public MessageSeverity getSeverity() {
        return severity;
    }

    /**
     * Returns the text of this message.
     * @return the text
     */
    public String getText() {
        return text;
    }

    @Override
    public String toString() {
        return String.format("%s: %s: %s: %s", source, id, severity, text);
    }
}
